src/dmd/func.d

```diff
@@ -477,93 +456,65 @@ extern (C++) class FuncDeclaration : Declaration
-            if (type.ty == Tfunction)
-            {
-                TypeFunction tf = cast(TypeFunction)type;
-                if (tf.isreturn)
-                    v2.storage_class |= STC.return_;
-                if (tf.isscope)
-                    v2.storage_class |= STC.scope_;
-                // if member function is marked 'inout', then this is 'return ref'
-                if (tf.iswild & 2)
-                    v2.storage_class |= STC.return_;
-            }
-            if (flags & FUNCFLAG.inferScope && !(v2.storage_class & STC.scope_))
-                v2.storage_class |= STC.maybescope;
-            v2.dsymbolSemantic(sc);
-            if (!sc.insert(v2))
-                assert(0);
-            v2.parent = this;
-            return HiddenParameters(v2, true);
+            vthis = null;
+            selectorParameter = null;
+            return;
         }
-        if (ad)
-        {
-            //printf("declareThis() %s\n", toChars());
-            Type thandle = ad.handleType();
-            assert(thandle);
-            thandle = thandle.addMod(type.mod);
-            thandle = thandle.addStorageClass(storage_class);
-            VarDeclaration v = new ThisDeclaration(loc, thandle);
-            v.storage_class |= STC.parameter;
-            if (thandle.ty == Tstruct)
-            {
-                v.storage_class |= STC.ref_;
-                // if member function is marked 'inout', then 'this' is 'return ref'
-                if (type.ty == Tfunction && (cast(TypeFunction)type).iswild & 2)
-                    v.storage_class |= STC.return_;
-            }
-            if (type.ty == Tfunction)
-            {
-                TypeFunction tf = cast(TypeFunction)type;
-                if (tf.isreturn)
-                    v.storage_class |= STC.return_;
-                if (tf.isscope)
-                    v.storage_class |= STC.scope_;
-            }
-            if (flags & FUNCFLAG.inferScope && !(v.storage_class & STC.scope_))
-                v.storage_class |= STC.maybescope;
 
-            v.dsymbolSemantic(sc);
-            if (!sc.insert(v))
-                assert(0);
-            v.parent = this;
-            return HiddenParameters(v, false, objc.createSelectorParameter(this, sc));
+        Type addModStc(Type t)
+        {
+            return t.addMod(type.mod).addStorageClass(storage_class);
         }
-        if (isNested())
+
+        if (isThis2 || isNested())
         {
             /* The 'this' for a nested function is the link to the
              * enclosing function's stack frame.
              * Note that nested functions and member functions are disjoint.
              */
-            VarDeclaration v = new VarDeclaration(loc, Type.tvoid.pointerTo(), Id.capture, null);
-            v.storage_class |= STC.parameter | STC.nodtor;
-            if (type.ty == Tfunction)
+            Type tthis = addModStc(isThis2 ?
+                                   Type.tvoidptr.sarrayOf(2).pointerTo() :
```

Comment:
Note to self, this type could be cached somewhere for easy comparison in the backend where `void*[2]*` can be replaced with a two field struct with the actual context types as its fields.